Implement styles

ddimaria opened the pull request:
Add style support for Calamine.

Basic api for getting styles
```rust
fn test_worksheet_style_iter() {
    let mut xlsx: Xlsx<_> = wb("styles.xlsx");
    let styles = xlsx.worksheet_style("Sheet 1").unwrap();

    for (row, styles) in styles.rows().enumerate() {
        for (col, style) in styles.iter().enumerate() {
            println!("row: {}, col: {}, style: {:?}", row, col, style);
        }
    }
}
```

Here is the relevant structs so far:
```rust
struct Style {
    font: Option<Font>,
    fill: Option<Fill>,
    borders: Option<Borders>,
    alignment: Option<Alignment>,
    number_format: Option<NumberFormat>,
    protection: Option<Protection>,
    style_id: Option<u32>,
}

struct Color {
    alpha: u8,
    red: u8,
    green: u8,
    blue: u8,
}

struct Borders {
    left: Border,
    right: Border,
    top: Border,
    bottom: Border,
    diagonal_down: Border,
    diagonal_up: Border,
}

struct Font {
    name: Option<String>,
    size: Option<f64>,
    weight: FontWeight,
    style: FontStyle,
    underline: UnderlineStyle,
    strikethrough: bool,
    color: Option<Color>,
    family: Option<String>,
}

struct Alignment {
    horizontal: HorizontalAlignment,
    vertical: VerticalAlignment,
    text_rotation: TextRotation,
    wrap_text: bool,
    indent: Option<u8>,
    shrink_to_fit: bool,
}

struct Fill {
    pattern: FillPattern,
    foreground_color: Option<Color>,
    background_color: Option<Color>,
}

struct NumberFormat {
    format_code: String,
    format_id: Option<u32>,
}

struct Protection {
    locked: bool,
    hidden: bool,
}
```

azarmadr commented:
Is the data and styles coming together?
If not, then would we have to read the excel twice to get data, once and then style?

azarmadr commented:
i got it, even formulae were being used similarly. Was able to satisfy my requirement with `worksheet_sytle` over a range.

Another doubt, why every item in the style is an option?
All the fonts, borders, and others are available as ids in cellxfs?

hope i am not bothering you.😊

ddimaria commented:
> Another doubt, why every item in the style is an option?

@azarmadr this is because the particular style may not exist for a given cell.

ddimaria commented:
@jmcnamara We're currently using this fork in production and it's working well.  What do I need to do to get this merged with Calamine?  Some things to discuss:
* Is the API agreeable?
* What type of documentation is needed?
* Add more testing?
* Should this be a compiler feature flag?

jmcnamara commented on examples/style_example.rs:
```diff
@@ -0,0 +1,92 @@
+// SPDX-License-Identifier: MIT
```
+1 for example programs.

jmcnamara commented on src/datatype.rs:
```diff
@@ -21,6 +22,67 @@ const EXCEL_1900_1904_DIFF: f64 = 1462.;
+    /// The cell value
+    pub value: Data,
+    /// The cell style
+    pub style: Option<Style>,
```
Strictly speaking the style is never optional in the Excel file formats. There is always an implicit 0 style that is the default and is defined in the `theme.xml` file. Note, I'm not necessarily saying that it shouldn't be an `Option<>`.

jmcnamara commented on src/datatype.rs:
```diff
@@ -21,6 +22,67 @@ const EXCEL_1900_1904_DIFF: f64 = 1462.;
+        self.style = None;
+    }
+
+    /// Check if the cell has style information
```
Again, an Excel cell will always have style information even if it is implicit for size reasons.

jmcnamara commented on src/lib.rs:
```diff
@@ -451,6 +457,39 @@ impl<T: CellType> Cell<T> {
+        }
+    }
+
+    /// Creates a new `Cell` instance with style information.
```
This level of documentation is good.

jmcnamara commented on src/lib.rs:
````diff
@@ -489,6 +528,86 @@ impl<T: CellType> Cell<T> {
+    /// assert!(cell.get_style().is_some());
+    /// ```
+    ///
+    pub fn set_style(&mut self, style: Style) {
````
Again, why is there a setter (and remove) method.

jmcnamara commented on src/xlsx/style_parser.rs:
```diff
@@ -0,0 +1,469 @@
+                // TODO: Handle indexed colors
+                return Ok(None);
+            }
+            _ => {}
```
You may need to consider the "Automatic" and "Default" color options.

jmcnamara commented on src/style.rs:
```diff
@@ -0,0 +1,768 @@
+    }
+
+    /// Check if the color is transparent
+    pub fn is_transparent(&self) -> bool {
```
Transparency is usually stored as a separate parameter outside the color. I don't think this would actually tell the user if the color was transparent.

jmcnamara commented on src/xlsx/style_parser.rs:
```diff
@@ -46,15 +57,28 @@ fn parse_color(attributes: &[Attribute]) -> Result<Option<Color>, XlsxError> {
     match s {
-        "bold" => FontWeight::Bold,
-        _ => FontWeight::Normal,
+        "bold" | "700" => FontWeight::Bold,
```
I think technically in the Xls format bold can have other weights but I haven't seen it in practice.

jmcnamara commented on src/style.rs:
```diff
@@ -603,6 +603,195 @@ impl Protection {
+    }
+
+    /// Get effective column width (custom or default)
+    pub fn get_effective_column_width(&self, column: u32) -> f64 {
```
This effective width is probably not the width that the user would expect and also may be dependent on the default font for the file.

See the following for the calculation in the other direction: https://github.com/jmcnamara/rust_xlsxwriter/blob/main/src/worksheet.rs#L19340-L19368

However, there may not be any simple way to return the correct value in all cases.

jmcnamara commented on src/style.rs:
```diff
@@ -603,6 +603,195 @@ impl Protection {
+    }
+
+    /// Get effective row height (custom or default)
+    pub fn get_effective_row_height(&self, row: u32) -> f64 {
```
Same comment for row height although the height calculation is simpler.

jmcnamara commented:
First review done. Some general comments:

* Please squash and or rebase WIP code and reverts into a small number of logical commits or one commit.
* Fix the clippy and or format warnings the are breaking the CI.
* Add a full stop/period at the end of all comments.
* Avoid /// comments for non-public non-doc comments. Use // instead.
* Post an explanation here for the setting functions in `CellData` and `Style`.
*

ddimaria commented:
> @ddimaria Any update on this?

@jmcnamara sorry for the delay, I got pulled into other projects.  My goal is to polish this PR and address feedback next week.  We've been using the fork in production, so the base of the code is working for us.  As a stretch goal I'd like to implement a better storage structure to account for repeating elements (RLE).

jmcnamara commented:
> My goal is to polish this PR and address feedback next week. We've been using the fork in production, so the base of the code is working for us. As a stretch goal I'd like to implement a better storage structure to account for repeating elements (RLE).

@ddimaria I would still like to merge this since it would fix a lot of other feature requests. I would avoid the stretch goal for now.

ddimaria commented on src/xlsx/cells_reader.rs:
```diff
@@ -134,6 +120,21 @@ where
+                    };
+
+                    if style_id < self.styles.len() {
+                        style = Some(self.styles[style_id].clone());
```
I'd like to get some perf number before implementing.  We've been using this branch in production and haven't noticed perf issues.

louisdewar replied:
Sorry for the delay.

Simple benchmark:
```rs
use std::env;
use calamine::{open_workbook_auto, Reader};

fn main() {
    let file_path = env::args().nth(1).expect("Please provide a file path");

    let mut workbook = open_workbook_auto(&file_path)
        .expect("Failed to open workbook");

    for sheet_name in workbook.sheet_names() {
        let range = workbook
            .worksheet_range(&sheet_name)
            .expect("Failed to read worksheet");

        for row in range.rows() {
            for cell in row {
                std::hint::black_box(cell);
            }
        }
    }
}
```

## Before
```
# warmup = 2
Benchmark 1: ./target/release/examples/parse_tmp ~/Downloads/DESRI_BASE.xlsm
  Time (mean ± σ):      1.285 s ±  0.005 s    [User: 1.206 s, System: 0.075 s]
  Range (min … max):    1.275 s …  1.292 s    10 runs
```
https://share.firefox.dev/4i0pUAo
7.2% of total extraction time is spent doing a specific style.clone().

## After removing .clones
I'd imagine we would have comparible performance between this and just storing the style ID for each cell.
```
$ git diff
diff --git a/src/xlsx/cells_reader.rs b/src/xlsx/cells_reader.rs
index b1b1168..e60dc2f 100644
--- a/src/xlsx/cells_reader.rs
+++ b/src/xlsx/cells_reader.rs
@@ -132,7 +132,7 @@ where
                     };

                     if style_id < self.styles.len() {
-                        style = Some(self.styles[style_id].clone());
+                        // style = Some(self.styles[style_id].clone());
                     }

                     loop {
@@ -211,7 +211,7 @@ where
                     };

                     if style_id < self.styles.len() {
-                        style = Some(self.styles[style_id].clone());
+                        // style = Some(self.styles[style_id].clone());
                     }

                     loop {
@@ -352,7 +352,8 @@ where
                     {
                         if let Ok(style_id) = atoi_simd::parse::<usize>(style_id_str) {
                             if style_id < self.styles.len() {
-                                self.styles[style_id].clone()
+                                Style::new()
+                                // self.styles[style_id].clone()
                             } else {
                                 Style::new()
                             }
```

```
# warmup = 2
Benchmark 1: ./target/release/examples/parse_tmp ~/Downloads/DESRI_BASE.xlsm
  Time (mean ± σ):      1.103 s ±  0.005 s    [User: 1.038 s, System: 0.061 s]
  Range (min … max):    1.093 s …  1.110 s    10 runs
```
https://share.firefox.dev/47WkCBo

About a 14% reduction in time.

## Notes

I assume the easiest way to get this speedup is to simply store style ID in the cells and expose the style table.

I haven't profiled memory, I'm not sure what the best test of this is but I imagine this would also show a reasonable impact assuming you were storing the entire workbook in memory e.g. doing a .collect() on the cell reader.

Also this benchmark could easily be biased, I just picked a large xlsx I was using at work, other workbooks might show something different.

My laptop is quite fast (m4 pro), I can imagine the different is a lot bigger on slower servers etc...

jmcnamara commented:
@ddimaria Any update on this. I would really like to get this upstream in the next few months.

See also: #597

louisdewar commented:
I'm also willing to help out on this if that would be useful

jmcnamara commented:
@louisdewar that would be helpful.

ddimaria commented:
@jmcnamara I need to implement conditional formatting and validations, is it OK to add them to this PR?

jmcnamara commented:
@ddimaria This PR is already very large. It would be best to get this stabilized first (is it suitable for review now?). 

You can create other PRs that are dependent on this on. Conditional formatting and Data Validation should be separate PRs.

Erik1000 commented:
> > @ddimaria This PR is already very large. It would be best to get this stabilized first (is it suitable for review now?).
> > You can create other PRs that are dependent on this on. Conditional formatting and Data Validation should be separate PRs.
> 
> That works for me. I just rebased my fork's commits onto upstream/master and squashed into a single commit. I'll do a final read through the code this morning and mark this ready for review.

is this PR now ready for review and can be merge? 👀

ddimaria commented:
> > > @ddimaria This PR is already very large. It would be best to get this stabilized first (is it suitable for review now?).
> > > You can create other PRs that are dependent on this on. Conditional formatting and Data Validation should be separate PRs.
> > 
> > 
> > That works for me. I just rebased my fork's commits onto upstream/master and squashed into a single commit. I'll do a final read through the code this morning and mark this ready for review.
> 
> is this PR now ready for review and can be merge? 👀

@Erik1000 I believe it is, though some conflicts just popped up.  Let me review those and CI and it will be ready to review/merge.

ddimaria commented:
@jmcnamara this PR is ready for review.  I'm staring conditional formatting in another branch.

jmcnamara commented:
> this PR is ready for review. I'm staring conditional formatting in another branch.

Ok. Fix the `cargo fmt` issues and if present, any clippy warnings. I think it would be best to resubmit this as a new PR. There are a lot of out of date comments on this PR.

Erik1000 commented:
@ddimaria will you submit a new PR or is there already an existing one?

ddimaria commented:
I'm back on this today.  Will resolve conflicts and start a new PR

ddimaria commented:
PR moved to https://github.com/tafia/calamine/pull/653, it OK to close this one.
